source/postprocess/crystal_preferred_orientation.cc

```diff
@@ -0,0 +1,816 @@
+
+              // write data
+              for (unsigned int grain_i = 0; grain_i < n_grains; ++grain_i)
+                {
+                  string_stream_content_draw_volume_weighting << id << " ";
+                  for (unsigned int property_i = 0; property_i < write_draw_volume_weighted_cpo.size(); ++property_i)
+                    {
+                      switch (write_draw_volume_weighted_cpo[property_i].second)
+                        {
+                          case Output::VolumeFraction:
+                            string_stream_content_draw_volume_weighting << volume_fractions_grains[write_draw_volume_weighted_cpo[property_i].first][grain_i] << " ";
+                            break;
+
+                          case Output::RotationMatrix:
+                            string_stream_content_draw_volume_weighting << weighted_rotation_matrices[write_draw_volume_weighted_cpo[property_i].first][grain_i] << " ";
+                            break;
+
+                          case Output::EulerAngles:
+                            Assert(compute_raw_euler_angles == true,
+                                   ExcMessage("Internal error: writing out raw Euler angles, without them being computed."));
+                            string_stream_content_draw_volume_weighting << weighted_euler_angles[write_draw_volume_weighted_cpo[property_i].first][grain_i][0] << " "
+                                                                        <<  weighted_euler_angles[write_draw_volume_weighted_cpo[property_i].first][grain_i][1] << " "
+                                                                        <<  weighted_euler_angles[write_draw_volume_weighted_cpo[property_i].first][grain_i][2] << " ";
+                            break;
+
+                          default:
+                            Assert(false, ExcMessage("Internal error: raw CPO postprocess case not found."));
+                            break;
+                        }
+                    }
+                  string_stream_content_draw_volume_weighting << std::endl;
+                }
+            }
+        }
+
+      std::string filename_master = particle_file_prefix_master + "." + Utilities::int_to_string(dealii::Utilities::MPI::this_mpi_process (MPI_COMM_WORLD),4) + ".dat";
+      std::string filename_raw = particle_file_prefix_content_raw + "." + Utilities::int_to_string(dealii::Utilities::MPI::this_mpi_process (MPI_COMM_WORLD),4) + ".dat";
+      std::string filename_draw_volume_weighting = particle_file_prefix_content_draw_volume_weighting + "." + Utilities::int_to_string(dealii::Utilities::MPI::this_mpi_process (MPI_COMM_WORLD),4) + ".dat";
+
+      std::string *file_contents_master = new std::string (string_stream_master.str());
+      std::string *file_contents_raw = new std::string (string_stream_content_raw.str());
+      std::string *file_contents_draw_volume_weighting = new std::string (string_stream_content_draw_volume_weighting.str());
```

Comment:
This works, but looks weird. I realize we do the same inside particles.cc. But at least use the better way of doing it from visualization.cc:570-575 where we use a std::unique_ptr instead of a raw pointer.